### PR TITLE
SONARJAVA-1532 support Java 9 changes in try-with-resources 

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 1]
```diff
@@ -25,6 +25,7 @@
 import org.sonar.check.RuleProperty;
 import org.sonar.java.matcher.MethodMatcher;
 import org.sonar.java.model.ExpressionUtils;
+import org.sonar.java.model.expression.IdentifierTreeImpl;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: MINOR Remove this unused import 'org.sonar.java.model.expression.IdentifierTreeImpl'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-frontend/src/main/java/org/sonar/plugins/java/api/tree/TryStatementTree.java [hunk 2]
```diff
@@ -45,10 +45,17 @@
...
 
   /**
+   * Java 9 allows resources to be declared by existing effectively final variables.
+   * This method will not return trees of resources declared by such variables, only resources declared with new variables will be returned (as it was until Java 9).
+   * Use {@link #resourceList()} to get all declared resources.
+   *
    * @since Java 1.7
    */
+  @Deprecated
   ListTree<VariableTree> resources();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/plugins/java/api/tree/TryStatementTree.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [UnclosedResourcesCheck.java#L51](https://github.com/SonarSource/sonar-java/blob/1c3bb458149142b1ce7fdf84ce7b522b3b1d5a90/java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java#L51): Remove this unused import 'org.sonar.plugins.java.api.tree.VariableTree'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: java-frontend/src/main/java/org/sonar/java/model/statement/TryStatementTreeImpl.java [hunk 3]
```diff
@@ -81,13 +85,33 @@ public TryStatementTreeImpl(
...
+        separator = separators.next();
+      }
+      if (resource.is(Kind.VARIABLE)) {
+        filteredResources.add(resource);
+        if (separator != null) {
+          filteredSeparators.add(separator);
+        }
+      }
+    }
+    return new ResourceListTreeImpl(filteredResources.build(), filteredSeparators.build());
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/model/statement/TryStatementTreeImpl.java]: I think a more elegant approach would be to directly modify the `ResourceListTreeImpl` in order to keep a distinct list of `VariableTree` on its side, which would be created at instantiation time as well. That way, we would keep that logic in one place and won't have to use a ListTree with unspecified type arguments.

File: java-frontend/src/main/java/org/sonar/plugins/java/api/tree/TryStatementTree.java [hunk 4]
```diff
@@ -45,10 +45,21 @@
...
+   *
+   * @deprecated since 4.11. Use {@link #resourceList()} to get all declared resources.
+   *
    * @since Java 1.7
    */
+  @Deprecated
   ListTree<VariableTree> resources();
 
+  /**
+   * @since Java 1.7, changed in Java 9 to allow also effectively final variables
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/plugins/java/api/tree/TryStatementTree.java]: wouldn't it be `@since Java 9` then?

Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/plugins/java/api/tree/TryStatementTree.java]: Improved in next commit.

**Pull request merged**